exercises/practice/list-ops/test_list_ops.c

```diff
@@ -1,14 +1,54 @@
-#include "test-framework/unity.h"
+#include <stdarg.h>
 #include "list_ops.h"
+#include "test-framework/unity.h"
 
 const int MAX_STRING_LEN = 100;
+list_t *list = NULL;
+list_t *list2 = NULL;
+list_t *actual = NULL;
+char *error_message = NULL;
 
 void setUp(void)
 {
 }
 
+static void call_delete_list(int count, ...)
+{
+   va_list ptr;
+   va_start(ptr, count);
+   for (int i = 0; i < count; i++) {
+      list_t **list = va_arg(ptr, list_t **);
+      delete_list(*list);
+      *list = NULL;
+   }
+   va_end(ptr);
+}
+
+static void release(int count, ...)
+{
+   va_list ptr;
+   va_start(ptr, count);
+   for (int i = 0; i < count; i++) {
+      if (i < 3) {
+         list_t **list = va_arg(ptr, list_t **);
+         if (*list) {
+            free(*list);
+            *list = NULL;
+         }
+      } else {
+         char **err_msg = va_arg(ptr, char **);
+         if (*err_msg) {
+            free(*err_msg);
+            *err_msg = NULL;
+         }
+      }
```

Comment:
This seems kind of convoluted. Why have a variadic function that's only called once and that has one argument handled in a special way?

Reply:
Some tests only have one list to delete; some have up to three. Also, the error_msg can leak memory. So we can have this

```c
void tearDown(void)
{
   release(4, &list, &list2, &actual, &error_message);
}
```

Instead of this

```c
static void release(list_t **list)
{
   if (*list)
   {
      free(*list);
      *list = NULL;
   }
}

void tearDown(void)
{
   release(&list);
   release(&list2);
   release(&actual);
   if (error_message)
   {
      free(error_message);
      error_message = NULL;
   }
}
```

But if you prefer the latter I can change it back.

Reply:
The part of this that feels clumsiest isn't the use of a variadic function, but the fact that one of the arguments is special in a non-obvious way. What is gained by including the message as one of the variadic arguments instead of as a positional parameter?

Reply:
The variadic function as it is now has all cleanup in one function vs. distributing it between itself and teardown. We could have the variadic take just the lists and still do `error_message` separately, like so

```c
static void release(int count, ...)
{
   va_list ptr;
   va_start(ptr, count);
   for (int i = 0; i < count; i++) {
      list_t **list = va_arg(ptr, list_t **);
      if (*list) {
         free(*list);
         *list = NULL;
      }
   }
   va_end(ptr);
}

void tearDown(void)
{
   release(3, &list, &list2, &actual);
   if (error_message) {
      free(error_message);
      error_message = NULL;
   }
}
```

I can implement whichever way you prefer.

Reply:
I think your last proposal is a good way to go.